File: LeetCode/2402-Meeting-Rooms-III.cpp
// 2402. Meeting Rooms III

class Solution {
public:
    int mostBooked(int n, vector<vector<int>>& meetings) {
        sort(meetings.begin(),meetings.end());

        priority_queue<int, vector<int>, greater<int>> unused;
        priority_queue<pair<long long,int>, vector<pair<long long,int>>, greater<pair<long long,int>>> used;
        for(int i = 0; i < n; i++) unused.push(i);

        vector<int> cnt(n, 0);
        for(auto m : meetings) {
            long long st = m[0];
            long long end = m[1];
            while(!used.empty() && used.top().first <= st) {
                unused.push(used.top().second);
                used.pop();
            }
            if(!unused.empty()) {
                cnt[unused.top()]++;
                used.push({end, unused.top()});
                unused.pop();
            }
            else {
                long long time = used.top().first;
                int room = used.top().second;
                used.pop();
                cnt[room]++;
                used.push({end - st + time, room});
            }
        }

        int mx = *max_element(cnt.begin(), cnt.end());
        for(int i = 0; i < n; i++) {
            if(cnt[i] == mx) return i;
        }
        return -1;
    }
};

File: LeetCode/2402-Meeting-Rooms-III.java
// 2402. Meeting Rooms III

class Solution {
    public int mostBooked(int n, int[][] meetings) {
        Arrays.sort(meetings, (a, b) -> a[0] - b[0]);

        PriorityQueue<Integer> rooms = new PriorityQueue<>();
        for (int i = 0; i < n; i++) rooms.add(i);

        int[] f = new int[n];
        PriorityQueue<long[]> freeAt = new PriorityQueue<>(
            (a, b) -> {
                if (a[0] != b[0]) return Long.compare(a[0], b[0]);
                return Long.compare(a[1], b[1]);
            }
        ); // [End, room]

        for (int[] m : meetings) {
            long start = m[0], end = m[1], delay = 0;
            while (!freeAt.isEmpty() && freeAt.peek()[0] <= start) {
                rooms.add((int)freeAt.poll()[1]);
            }

            // if there are no any free rooms yet then take the first one and add that delay
            if (rooms.isEmpty()) {
                delay = freeAt.peek()[0] - start;
                rooms.add((int)freeAt.poll()[1]);
            }

            long room = rooms.poll();
            f[(int)room]++;
            freeAt.add(new long[]{end + delay, room});
        }

        int mx = Integer.MIN_VALUE, ans = -1;
        for (int i = 0; i < n; i++) {
            if (f[i] > mx) {
                mx = f[i];
                ans = i;
            }
        }
        return ans;
    }
}
